From: Thomas Graf <tgraf@redhat.com>
Date: Thu, 23 Jun 2011 11:34:20 -0400
Subject: [net] inet_diag: fix inet_diag_bc_audit data validation
Message-id: <20110623113419.GA23009@lsx.localdomain>
Patchwork-id: 36921
O-Subject: [RHEL5.7 PATCH CVE-2011-2213] inet_diag: fix inet_diag_bc_audit() (BZ714539)
Bugzilla: 714539
CVE: CVE-2011-2213
RH-Acked-by: Petr Matousek <pmatouse@redhat.com>
RH-Acked-by: Neil Horman <nhorman@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>

Fixes inet_diag_bc_audit() to correctly validate the user data
preventing the possibility of triggering an infinite loop.

Fixes BZ714539 and CVE-2011-2213

Brew:
https://brewweb.devel.redhat.com/taskinfo?taskID=3423840

Upstream:
commit eeb1497277d6b1a0a34ed36b97e18f2bd7d6de0d
Author: Eric Dumazet <eric.dumazet@gmail.com>
Date:   Fri Jun 17 16:25:39 2011 -0400

    inet_diag: fix inet_diag_bc_audit()

    A malicious user or buggy application can inject code and trigger an
    infinite loop in inet_diag_bc_audit()

    Also make sure each instruction is aligned on 4 bytes boundary, to avoid
    unaligned accesses.

    Reported-by: Dan Rosenberg <drosenberg@vsecurity.com>
    Signed-off-by: Eric Dumazet <eric.dumazet@gmail.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv4/inet_diag.c b/net/ipv4/inet_diag.c
index dd43a67..f84369b 100644
--- a/net/ipv4/inet_diag.c
+++ b/net/ipv4/inet_diag.c
@@ -407,7 +407,7 @@ static int valid_cc(const void *bc, int len, int cc)
 			return 0;
 		if (cc == len)
 			return 1;
-		if (op->yes < 4)
+		if (op->yes < 4 || op->yes & 3)
 			return 0;
 		len -= op->yes;
 		bc  += op->yes;
@@ -417,11 +417,11 @@ static int valid_cc(const void *bc, int len, int cc)
 
 static int inet_diag_bc_audit(const void *bytecode, int bytecode_len)
 {
-	const unsigned char *bc = bytecode;
+	const void *bc = bytecode;
 	int  len = bytecode_len;
 
 	while (len > 0) {
-		struct inet_diag_bc_op *op = (struct inet_diag_bc_op *)bc;
+		const struct inet_diag_bc_op *op = bc;
 
 //printk("BC: %d %d %d {%d} / %d\n", op->code, op->yes, op->no, op[1].no, len);
 		switch (op->code) {
@@ -432,22 +432,20 @@ static int inet_diag_bc_audit(const void *bytecode, int bytecode_len)
 		case INET_DIAG_BC_S_LE:
 		case INET_DIAG_BC_D_GE:
 		case INET_DIAG_BC_D_LE:
-			if (op->yes < 4 || op->yes > len + 4)
-				return -EINVAL;
 		case INET_DIAG_BC_JMP:
-			if (op->no < 4 || op->no > len + 4)
+			if (op->no < 4 || op->no > len + 4 || op->no & 3)
 				return -EINVAL;
 			if (op->no < len &&
 			    !valid_cc(bytecode, bytecode_len, len - op->no))
 				return -EINVAL;
 			break;
 		case INET_DIAG_BC_NOP:
-			if (op->yes < 4 || op->yes > len + 4)
-				return -EINVAL;
 			break;
 		default:
 			return -EINVAL;
 		}
+		if (op->yes < 4 || op->yes > len + 4 || op->yes & 3)
+			return -EINVAL;
 		bc  += op->yes;
 		len -= op->yes;
 	}
